### PR TITLE
README update, prevent broken pipe error

### PR DESCRIPTION
**Pull request by lzap**
Review of README because Red Hat doco team had a lot of questions when
processing this and I decided to fix this. Some comments inline.

Also this fixes EPIPE which can occur when script closes STDIN.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -24,7 +24,7 @@ RPM users can install the "ruby193-rubygem-foreman_hooks" or
 Hooks are stored in `/usr/share/foreman/config/hooks` (`~foreman/config/hooks`)
 with a subdirectory for the object, then a subdirectory for the event name.
 
-    ~foreman/config/hooks/$OBJECT/$EVENT/$HOOK_SCRIPT
+    ~foreman/config/hooks/[OBJECT]/[EVENT]/[HOOK_SCRIPT]
```
Comment (lzap) [inline README.md]: This looked like shell variables, I think brackets are better.

File: README.md [hunk 2]
```diff
@@ -33,33 +33,97 @@ Examples:
...
-SELinux is turned on by default and Foreman is running in confined mode. Make
-sure that hook scripts has the correct context (`foreman_hook_t` on
-RHEL7+/Fedora 19+ or `bin_t` on RHEL6):
-
-    restorecon -RvF /usr/share/foreman/config/hooks
-
-Also keep in mind that the script is running confined, therefore some actions
-might be denied by SELinux. Check audit.log and use audit2allow and other
-tools when writing scripts.
-
```
Comment (lzap) [inline README.md]: Moved SELinux below.

File: README.md [hunk 3]
```diff
@@ -33,33 +33,97 @@ Examples:
...
+* `taxonomy`
+* `template`
+* `user`
+* `user_role`
+* `usergroup`
+
+To generate list of *all* possible models, issue the following command:
+
+    # foreman-rake console
+    > ActiveRecord::Base.descendants.collect(&:name).collect(&:underscore).sort
```
Comment (lzap) [inline README.md]: I used this command to generate the list. I think it won't hurt us to explicitly list it all here.

File: README.md [hunk 4]
```diff
@@ -122,6 +190,25 @@ to rollback its action - in this case the first argument will change as
 appropriate, so must be obeyed by the script (e.g. a "create" hook will be
 called with "destroy" if it has to be rolled back later).
 
+## Transactions
+
+Most hooks are triggered during database transaction. This can cause deadlocks
+when hook scripts emits database updates via Foreman CLI or API. It is
+recommended to avoid this behavior and write a Foreman Plugin instead.
```
Comment (lzap) [inline README.md]: I had a downstream BZ when customer was actually updating hosts via hooks encountering errors on load.

File: README.md [hunk 5]
```diff
@@ -33,33 +33,97 @@ Examples:
...
-tools when writing scripts.
-
 ## Objects / Models
 
 Every object (or model in Rails terms) in Foreman can have hooks.  Check
-`~foreman/app/models` for the full list, but these are the interesting ones:
-
-* `host/managed` (or `host` in Foreman 1.1)
+`~foreman/app/models` for the full list, but these are the interesting ones
+from Foreman 1.11 Core or some its plugins:
```
Comment (domcleal) [inline README.md]: Please remove this list, I don't want to maintain it here and other plugins shouldn't be listed.

File: README.md [hunk 6]
```diff
@@ -67,10 +131,14 @@ To add hooks to these, use these event names:
 * `update`
 * `destroy`
 
+Orchestration hooks can be given a priority (see below), therefore it is
+possible to order them before or after built-in orchestration steps (before
+DNS record is deployed for example).
```
Comment (domcleal) [inline README.md]: s/deployed/created/ for consistency in terminology.

File: README.md [hunk 7]
```diff
@@ -67,10 +131,14 @@ To add hooks to these, use these event names:
...
+
 ## Rails events
 
-For hooks on anything apart from hosts (which support orchestration, as above)
-then the standard Rails events will be needed.  These are the most interesting
+For hooks on anything apart from hosts or NICs (which support orchestration, as above)
```
Comment (domcleal) [inline README.md]: wrap text

File: README.md [hunk 8]
```diff
@@ -122,6 +190,25 @@ to rollback its action - in this case the first argument will change as
 appropriate, so must be obeyed by the script (e.g. a "create" hook will be
 called with "destroy" if it has to be rolled back later).
 
+## Transactions
+
+Most hooks are triggered during database transaction. This can cause deadlocks
+when hook scripts emits database updates via Foreman CLI or API. It is
```
Comment (domcleal) [inline README.md]: conflicting updates

File: README.md (hunk 4 above)
Comment (domcleal) [inline README.md]: lower case "Plugin", and I don't think that's good advice, perhaps better to remove it.

File: lib/foreman_hooks/util.rb [hunk 9]
```diff
@@ -44,8 +44,12 @@ def exec_hook_int(stdin_data, *args)
       Open3.capture2e(*args.push(:stdin_data => stdin_data))
     else  # 1.8
       Open3.popen3(*args) do |stdin,stdout,stderr|
-        stdin.write(stdin_data)
-        stdin.close
+        begin
+          stdin.write(stdin_data)
+          stdin.close
+        rescue Errno::EPIPE
+          logger.debug "Input data not passed"
```
Comment (domcleal) [inline lib/foreman_hooks/util.rb]: This message is too short and contextless to be useful I think, it should give more information about what "data" is and what it's "passing" to - it won't be clear that this is from foreman_hooks.

File: README.md (hunk 4 above)
Comment (lzap) [inline README.md]: Why this is not a good advice?

File: README.md (hunk 5 above)
Comment (lzap) [inline README.md]: What you have right now is insufficient, after reading the README you have no idea what you can do with hooks. I've shortened the list to four, but I think it's better to have outdated documentation than no documentation.

File: README.md [hunk 10]
```diff
@@ -33,33 +33,97 @@ Examples:
...
+    # foreman-rake console
+    > ActiveRecord::Base.descendants.collect(&:name).collect(&:underscore).sort
 
 ## Orchestration events
 
 Foreman supports orchestration tasks for hosts and NICs (each network
 interface) which happen when the object is created, updated and destroyed.
 These tasks are shown to the user in the UI and if they fail, will
-automatically trigger a rollback of the action.
+automatically trigger a rollback (usually `destroy`) of the action.
```
Comment (lzap) [inline README.md]: Made it clearer.

**Comment by lzap** [comment]
Amended the changes.

File: README.md [hunk 11]
```diff
@@ -33,45 +33,45 @@ Examples:
...
-might be denied by SELinux. Check audit.log and use audit2allow and other
-tools when writing scripts.
-
 ## Objects / Models
 
 Every object (or model in Rails terms) in Foreman can have hooks.  Check
 `~foreman/app/models` for the full list, but these are the interesting ones:
 
 * `host/managed` (or `host` in Foreman 1.1)
-* `report`
```
Comment (domcleal) [inline README.md]: Why's this removed?

File: README.md [hunk 12]
```diff
@@ -33,45 +33,45 @@ Examples:
...
+* `nic/managed`
+* `hostgroup`
+* `user`
+
+To generate list of *all* possible models, issue the following command:
```
Comment (domcleal) [inline README.md]: `generate a list`

**Comment by domcleal** [comment]
> Also this fixes EPIPE which can occur when script closes STDIN.

If you could split documentation updates from bug fixes, that'd be appreciated.

**Comment by lzap** [comment]
> If you could split documentation updates from bug fixes, that'd be appreciated.

Sure, fixed both your comments.

-- 
Later,
 Lukas #lzap Zapletal

**Comment by domcleal** [comment]
Merged, thanks @lzap.

**Pull request merged**